orchagent/copporch.cpp

```diff
@@ -0,0 +1,479 @@
+                            return task_process_status::task_failed;
+                        }                    
+                    }
+                }
+            }
+            SWSS_LOG_DEBUG("Applying trap group attributes");
+            for(sai_uint32_t ind = 0; ind < trap_gr_attribs.size(); ind++)
+            {
+                auto trap_gr_attr = trap_gr_attribs[ind];
+                SWSS_LOG_DEBUG("Applying trap group attribute:%d", trap_gr_attr.id);
+                if (SAI_STATUS_SUCCESS != (sai_status = sai_hostif_api->set_trap_group_attribute(m_trap_map[key], &trap_gr_attr)))
+                {
+                    SWSS_LOG_ERROR("Failed to apply attribute:%d to trap group:%llx, name:%s, error:%d\n", trap_gr_attr.id, m_trap_map[key], key.c_str(), sai_status);
+                    return task_process_status::task_failed;
+                }
+            }                
+        }
+        else
+        {
+            SWSS_LOG_DEBUG("Creating new trap_group object:%s", key.c_str());
+            sai_object_id_t new_trap;
+            if (SAI_STATUS_SUCCESS != (sai_status = sai_hostif_api->create_hostif_trap_group(&new_trap, trap_gr_attribs.size(), trap_gr_attribs.data())))
+            {
+                SWSS_LOG_ERROR("Failed to create new trap_group with name:%s", key.c_str());
+                return task_process_status::task_failed;
+            }
+            SWSS_LOG_DEBUG("Created new trap_group:%llx with name:%s", new_trap, key.c_str());
+            m_trap_map[key] = new_trap;
+
+            if(!policer_attribs.empty())
+            {
+                if (SAI_STATUS_SUCCESS != (sai_status = sai_policer_api->create_policer(&policer_id, policer_attribs.size(), policer_attribs.data())))
+                {
+                    SWSS_LOG_ERROR("Failed to create policer for new trap_group:%llx, name:%s, error:%d", new_trap, key.c_str(), sai_status);
+                    return task_process_status::task_failed;
+                }
+                SWSS_LOG_DEBUG("Created new policer:%llx", policer_id);
+                attr.id = SAI_HOSTIF_TRAP_GROUP_ATTR_POLICER;
+                attr.value.oid = policer_id;
+                if (SAI_STATUS_SUCCESS != (sai_status = sai_hostif_api->set_trap_group_attribute(new_trap, &attr))) {
+                    SWSS_LOG_ERROR("Failed to bind policer:%llx to trap gorup:%llx, name:%s, error:%d\n", policer_id, new_trap, key.c_str(), sai_status);
+                    return task_process_status::task_failed;
+                }
+                SWSS_LOG_DEBUG("Bound policer to the trap group");
+            }
+        }
+
+        //
+        // The channel is always the NETDEV device.
+        //
+        attr.id = SAI_HOSTIF_TRAP_ATTR_TRAP_CHANNEL;
+        attr.value.s32 = SAI_HOSTIF_TRAP_CHANNEL_NETDEV;
+        trap_id_attribs.push_back(attr);
```

Comment:
I feel if you first create an internal trap object, and then translate the internal trap object into sai trap object, it maybe easier to read.

---
In reply to: [73782200](https://github.com/Azure/sonic-swss/pull/57#discussion_r73782200) [](ancestors = 73782200)